### PR TITLE
Add Portuguese keymap and sendstring lookup tables

### PR DESCRIPTION
**Pull request by luismbo**
## Description

Portuguese keymap. I haven't tested it since I don't have a QMK-capable keyboard yet and it didn't seem straightforward to write a test using send_keyboard_mock().

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* none

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/keymap_extras/keymap_portuguese.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+#define PT_HASH S(PT_3)    // #
+#define PT_DLR  S(PT_4)    // $
+#define PT_PERC S(PT_5)    // %
+#define PT_AMPR S(PT_6)    // &
+#define PT_SLSH S(PT_7)    // /
+#define PT_LPRN S(PT_8)    // (
+#define PT_RPRN S(PT_9)    // )
+#define PT_EQL  S(PT_0)    // =
+#define PT_QUES S(PT_QUOT) // ?
+#define PT_RADQ S(PT_LADQ) // »
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_portuguese.h]:
```suggestion
#define PT_RDAQ S(PT_LDAQ) // »
```

File: quantum/keymap_extras/keymap_portuguese.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+#define PT_AT   ALGR(PT_2)    // @
+#define PT_PND  ALGR(PT_3)    // £
+#define PT_SECT ALGR(PT_4)    // §
+#define PT_EURO ALGR(PT_5)    // €
+#define PT_LCBR ALGR(PT_7)    // {
+#define PT_LBRC ALGR(PT_8)    // [
+#define PT_RBRC ALGR(PT_9)    // ]
+#define PT_RCBR ALGR(PT_0)    // }
+// Row 2
+#define PT_EUR2 ALGR(PT_E)    // €
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_portuguese.h]:
```suggestion
```

**Comment by luismbo** [comment]
Accepted @fauxpark's edit suggestions, thanks. Let me know if you need me to squash the commits on my side.

File: quantum/keymap_extras/keymap_portuguese.h [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+#define PT_RABK S(PT_LABK) // >
+#define PT_SCLN S(PT_COMM) // ;
+#define PT_COLN S(PT_DOT)  // :
+#define PT_UNDS S(PT_MINS) // _
+
+/* AltGr symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┐
+ * │   │   │ @ │ £ │ § │ € │   │ { │ [ │ ] │ } │   │   │       │
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┤
+ * │     │   │   │ € │   │   │   │   │   │   │   │ ¨ │   │     │
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_portuguese.h]:
```suggestion
 * │     │   │   │   │   │   │   │   │   │   │   │ ¨ │   │     │
```

**Comment by noroadsleft** [comment]
I think 63ef0fd3 and 33ab1c0 should be reverted. Windows actually puts the Euro currency symbol on both AltGr+5 and AltGr+E:

image

**Comment by fauxpark** [comment]
Yep, but we really only need the one. It's not like some other layouts which have both a dead and "normal" version of certain accents, and I think it'd be preferable to avoid the confusion over whether there is any real difference between the two.

**Comment by luismbo** [comment]
The only reason that occurs to me is some OS that binds AltGr+E but not AltGr+5. BTW, I have two portuguese keyboards right here and one of them labels the E key with € and the other one labels the 5 key. Neither labels both. 🤷‍♂️

**Comment by noroadsleft** [comment]
Just checked on my Linux Mint netbook and `€` is on AltGr+E. AltGr+5 is `¢`.

**Comment by fauxpark** [comment]
Are you sure you're not looking at the Brazilian layout?
https://en.wikipedia.org/wiki/Portuguese_keyboard_layout

However, it does seem that `€` should be on AltGr+`E` instead.

**Comment by noroadsleft** [comment]
Oops, you're right, I forgot that Brazil and Portugal don't use the same layout or mapping.

Tried the same system with Portugal's layout. AltGr+E is `€`; AltGr+5 is `½`.

**Comment by fauxpark** [comment]
It probably warrants a separate keymap for Linux, like there already is for a couple layouts for macOS - but I have no idea if it's consistent across most/all distros. It may be a losing battle to try and support anything other than Windows, macOS, and, I dunno, Ubuntu?

**Comment by noroadsleft** [comment]
Linux Mint is based on Ubuntu, which itself is based on Debian, so I'd imagine they're likely all the same.

File: quantum/keymap_extras/keymap_portuguese.h [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+#define PT_RABK S(PT_LABK) // >
+#define PT_SCLN S(PT_COMM) // ;
+#define PT_COLN S(PT_DOT)  // :
+#define PT_UNDS S(PT_MINS) // _
+
+/* AltGr symbols
+ * ┌───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───┬───────┐
+ * │   │   │ @ │ £ │ § │ € │   │ { │ [ │ ] │ } │   │   │       │
+ * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┤
+ * │     │   │   │   │   │   │   │   │   │   │   │ ¨ │   │     │
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_portuguese.h]:
```suggestion
 * │   │   │ @ │ £ │ § │   │   │ { │ [ │ ] │ } │   │   │       │
 * ├───┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─────┤
 * │     │   │   │ € │   │   │   │   │   │   │   │ ¨ │   │     │
```

File: quantum/keymap_extras/keymap_portuguese.h [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+ * │    │   │   │   │   │   │   │   │   │   │   │   │          │
+ * ├────┼───┴┬──┴─┬─┴───┴───┴───┴───┴───┴──┬┴───┼───┴┬────┬────┤
+ * │    │    │    │                        │    │    │    │    │
+ * └────┴────┴────┴────────────────────────┴────┴────┴────┴────┘
+ */
+// Row 1
+#define PT_AT   ALGR(PT_2)    // @
+#define PT_PND  ALGR(PT_3)    // £
+#define PT_SECT ALGR(PT_4)    // §
+#define PT_EURO ALGR(PT_5)    // €
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_portuguese.h]:
```suggestion
```

File: quantum/keymap_extras/keymap_portuguese.h [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+#define PT_LCBR ALGR(PT_7)    // {
+#define PT_LBRC ALGR(PT_8)    // [
+#define PT_RBRC ALGR(PT_9)    // ]
+#define PT_RCBR ALGR(PT_0)    // }
+// Row 2
+#define PT_DIAE ALGR(PT_PLUS) // ¨ (dead)
```
Comment (fauxpark) [inline quantum/keymap_extras/keymap_portuguese.h]:
```suggestion
#define PT_DIAE ALGR(PT_PLUS) // ¨ (dead)
#define PT_EURO ALGR(PT_E)    // €
```

**Pull request merged**